vidpub/__main__.py

```diff
@@ -96,81 +94,209 @@ def get_match_ratio(session: Session, path: pathlib.Path) -> float:
     return fuzzywuzzy.fuzz.ratio(session.title, path.stem)
 
 
-def choose_video(session: Session) -> pathlib.Path:
+def choose_video(session: Session, video_paths: list) -> pathlib.Path:
     """Look through the file list and choose the one that "looks most like it".
     """
-    score, match = max((get_match_ratio(session, p), p) for p in VIDEO_PATHS)
+    score, match = max((get_match_ratio(session, p), p) for p in video_paths)
     if score < 70:
         raise ValueError("no match")
     return match
 
 
-def build_youtube():
+def build_youtube(action: str):
     from apiclient.discovery import build
     from google_auth_oauthlib.flow import InstalledAppFlow
 
-    flow = InstalledAppFlow.from_client_secrets_file(
-        os.environ["OAUTH2_CLIENT_SECRET"], scopes=[YOUTUBE_UPLOAD_SCOPE]
-    )
-    credentials = flow.run_console()
-    return build("youtube", "v3", credentials=credentials)
+    if action == 'upload':
+        flow = InstalledAppFlow.from_client_secrets_file(
+            os.environ["OAUTH2_CLIENT_SECRET"], scopes=[YOUTUBE_UPLOAD_SCOPE]
+        )
+        credentials = flow.run_console()
+        return build("youtube", "v3", credentials=credentials)
+
+    elif action == 'generate':
+        return build("youtube", "v3", developerKey=YOUTUBE_API_KEY)
+
+    else:
+        raise ValueError(f"Does not supported action = {action}")
+
+def compute_date(first_date: datetime.date, delta_days=0):
+    conference_date = first_date + datetime.timedelta(days=delta_days)
+
+    return str(conference_date)
+
+
+def extract_info(description: str):
 
+    speaker = []
+    recorded_day = ''
 
-def parse_args(argv):
-    parser = argparse.ArgumentParser()
-    parser.add_argument(
-        "--upload", action="store_true", help="Actually upload"
-    )
-    return parser.parse_args(argv)
+    lines = description.split('\n')
 
+    for line in lines:
+        if line.strip().lower().startswith('day'):
+            day = int(line.strip().lower().split(',')[0].strip().split(' ')[1])
+            recorded_day = compute_date(FIRST_DATE, delta_days=day-1)
+        elif line.strip().lower().startswith('speaker'):
+            speaker.append(line.strip().split(':')[1].strip())
+        else:
+            pass
 
-def main(argv=None):
-    options = parse_args(argv)
+    return speaker, recorded_day
 
-    if options.upload:
-        youtube = build_youtube()
+@click.command()
+@click.option('-a', '--action', required=True, help='Actions to perform: upload or generate', \
+                type=click.Choice(['upload', 'generate'], case_sensitive=False))
+@click.option('-o', '--output_dir', default='./videos', help='Output video information path')
 
-    source = ConferenceInfoSource(
-        requests.get(os.environ["URL"]).json(),
-        Conference(CONFERENCE_NAME, FIRST_DATE, TIMEZONE_TAIPEI),
-    )
+def main(action, output_dir):
```

Comment:
IMO, I would like to keep the main function clean by separating two types of action as functions from the main function. The benefits of this way will keep the original uploading codes as usual and only changing the signature and the new generating flow will look clean since the dependency will be reduced.

Reply:
I've updated the code based on the suggestion of separating two different actions into two functions.